### ISSUE TITLE
Event not triggering

### ISSUE THREAD

**Issue by kim0** [flexxui/flexx#523]
This is more of a question, than a bug. I'm sure I'm doing something wrong here. What I'm trying to achieve is, have a simple UI where the user can add multiple servers. Each with (IP, username and password). The user can add/delete servers. Then click "save" which saves in the backend (ideally to DB, but for now to a json file!)
My code is at in this [gist](https://gist.github.com/kim0/a0c74fb865fcff87a7ba048843398098)

Got a few questions:
* Is storing the list of servers on the Python side, then having the JS side react to changes on that and draw itself, consider best practice here?
* My main problem here is that the `servers_changed` event is not triggering on the JS side!
* Also strangely, when I use chrome's JS console to view `flexx.s1.instances.Samourai_app_1.servers` I find that it contains the contents of `servers.txt` but repeated twice (4 items!). No idea why really
* Finally, I'll probably need to add user authentication so that each user edits/saves his own list of servers. What's the simplest way that this can be achieved

Really thanks for Flexx. I could not have imagined to be working on a similar project without flexx :)

**Comment by almarklein** [comment]
> Is storing the list of servers on the Python side, then having the JS side react to changes on that and draw itself, consider best practice here?

Yes, that seems the most reasonable approach here, because you're loading from / saving to the file system. You could also consider keeping the list of servers in JS, e.g. when the property changes very fast due to user interaction, and then send to Py only upon saving, but I don't think this applies here.

> My main problem here is that the servers_changed event is not triggering on the JS side!

That's because you're listening to the "servers_changed" event on the UI object, and its emitted from the `Samourai_app` component. Try `connect('root.servers_changed')`, or simply `connect('root.servers')` should do as well.

> Also strangely, when I use chrome's JS console to view flexx.s1.instances.Samourai_app_1.servers I find that it contains the contents of servers.txt but repeated twice (4 items!). No idea why really

That's odd. Maybe this is related to using "insert" with the mutation? Try just setting it (`self._mutate_servers(servers_json)`).

> Finally, I'll probably need to add user authentication so that each user edits/saves his own list of servers. What's the simplest way that this can be achieved

There are multiple ways to approach Authentication. Flexx does not (yet?) provide tools for this. You could keep track of users yourself using a database and have them login. Or make use of OAuth with Github or Google. Or make use of a service like e.g. [Auth0](https://auth0.com/).

**Comment by kim0** [comment]
@almarklein Thanks a lot for your help, while I try to learn more about flexx! I hope you don't mind me using this thread for further questions :)

I'm now keeping the state on the py side, the event is now firing correctly (Thx!) and _mutate_servers works as it should. The updated code is at [here](https://gist.github.com/kim0/be202124058e10285f17859a21946ad8)

I think what I'm trying to do is pretty standard (load data from server, user edits in UI, then save back to server). However, things are not going smoothly for me

1. Loading the data server side, into a python data structure works great. The UI gets notified of the event and draws itself. The part is awesome!
1. But, when the user adds/edits a new server. I find that I have to hunt around the [web UI properties](https://gist.github.com/kim0/be202124058e10285f17859a21946ad8#file-sam-py-L42) gathering text properties, packing them into JS objects. Then throwing them over the wall to the [py side](https://gist.github.com/kim0/be202124058e10285f17859a21946ad8#file-sam-py-L93), where I again have to unpack that server list into a proper py data structure.

While 1 works great, 2 feels like too much busy work. Hunting around the UI is not fun, and constantly converting between JS objects and py types is also not fun. I know I'm probably not handling this the optimal way. What's a good way to make what I want as smooth as possible.

**Comment by almarklein** [comment]
Well, *somewhere* you should go from a dict-description of a server to the ui-specific representation. You're now doing that step at the server. I'd recommend doing that at the client. Maybe in `Server_entry`. You could even consider not having these 3 different properties, but one `info` property and populate the ui from that.
